ql/experimental/asian/analytic_discr_arith_av_price.cpp

```diff
@@ -0,0 +1,126 @@
+/*
+ Copyright (C) 2021 Skandinaviska Enskilda Banken AB (publ)
+
+ This file is part of QuantLib, a free-software/open-source library
+ for financial quantitative analysts and developers - http://quantlib.org/
+
+ QuantLib is free software: you can redistribute it and/or modify it
+ under the terms of the QuantLib license.  You should have received a
+ copy of the license along with this program; if not, please email
+ <quantlib-dev@lists.sf.net>. The license is also available online at
+ <http://quantlib.org/license.shtml>.
+
+ This program is distributed in the hope that it will be useful, but WITHOUT
+ ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ FOR A PARTICULAR PURPOSE.  See the license for more details.
+*/
+
+#include <ql/exercise.hpp>
+#include <ql/experimental/asian/analytic_discr_arith_av_price.hpp>
+#include <ql/pricingengines/blackformula.hpp>
+
+using namespace QuantLib;
+
+void AnalyticDiscreteArithmeticAveragePriceAsianEngine::calculate() const {
+
+    // Enforce a few required things
+    QL_REQUIRE(arguments_.exercise->type() == Exercise::European, "not a European Option");
```

Comment:
So, despite the option being "Asian", its exercise is still "European" (as opposed to, say, American or Bermudan), meaning that it can only be triggered at the sole expiry date. Hence the check for European exercise, only, here.